src/squlearn/qrc/base_qrc.py

```diff
@@ -0,0 +1,333 @@
+import numpy as np
+from typing import Union, List
+from abc import abstractmethod, ABC
+
+from sklearn.base import BaseEstimator
+from qiskit.quantum_info.random import random_pauli_list
+
+from ..observables.observable_base import ObservableBase
+from ..observables import CustomObservable, SinglePauli
+from ..encoding_circuit.encoding_circuit_base import EncodingCircuitBase
+from ..util import Executor
+from ..qnn.lowlevel_qnn import LowLevelQNN
+
+
+class BaseQRC(BaseEstimator, ABC):
+    """Base class for Quantum Reservoir Computing (QRC) models.
+
+    Args:
+        encoding_circuit (EncodingCircuitBase): The encoding circuit to use for encoding the data
+                                                into the reservoir.
+        executor (Executor): Executor instance
+        ml_model (str): The classical machine learning model to use (default: linear), possible
+                        values are ``"mlp"``, ``"linear"``, and ``"kernel"``.
+                        Implementation depends on the child.
+        ml_model_options (dict): The options for the machine learning model
+                                 (default options of the sklearn model are used if None)
+        operators (Union[ObservableBase, list[ObservableBase], str]): Strategy for generating the
+                                                                      operators used to measure the
+                                                                      quantum reservoir. Possible values are:
+
+                                                                        * ``"random_paulis"`` generates random Pauli operators (default).
+                                                                        * ``"single_paulis"`` generates single qubit Pauli operators.
+
+                                                                      Alternatively, a list of ObservableBase objects can be provided.
+        num_operators (int): The number of random Pauli operators to generate for
+                             ``"operators = random_paulis"`` (default: 100).
+        operator_seed (int): The seed for the random operator generation for
+                             ``"operators = random_paulis"`` (default: 0).
+        param_ini (Union[np.ndarray, None]): The parameters for the encoding circuit (optional).
+        param_op_ini (Union[np.ndarray, None]): The initial parameters for the operators (optional).
+        parameter_seed (Union[int, None]): The seed for the initial parameter generation
+                                           (if no parameters are given).
+        caching (bool): Whether to cache the results of the evaluated expectation values.
+    """
```

Comment:
Please only indent once in a new line, this is not readable. Also i adapted the docstrings a bit for clarity.
```suggestion
"""Base class for Quantum Reservoir Computing (QRC) models.

    Args:
        encoding_circuit (EncodingCircuitBase): The encoding circuit to use for encoding the data
            into the reservoir.
        executor (Executor): Executor instance
        ml_model (str): The classical machine learning model to use (default: linear), possible
            values are ``"mlp"``, ``"linear"``, and ``"kernel"``. Implementation depends on the
            child.
        ml_model_options (dict): The options for the machine learning model. Default options of the
            sklearn model are used if None.
        operators (Union[ObservableBase, list[ObservableBase], str]): Strategy for generating the
            operators used to measure the quantum reservoir. Possible values are:
                * ``"random_paulis"`` generates random Pauli operators (default).
                * ``"single_paulis"`` generates single qubit Pauli operators.
            Alternatively, a list of ObservableBase objects can be provided.
        num_operators (int): The number of random Pauli operators to generate for
            ``"operators = random_paulis"`` (default: 100).
        operator_seed (int): The seed for the random operator generation for
            ``"operators = random_paulis"`` (default: 0).
        param_ini (Union[np.ndarray, None]): The parameters for the encoding circuit.
        param_op_ini (Union[np.ndarray, None]): The initial parameters for the operators.
        parameter_seed (Union[int, None]): The seed for the initial parameter generation if no
            parameters are given.
        caching (bool): Whether to cache the results of the evaluated expectation values.
    """
```